src/main/java/org/gaul/s3proxy/junit/S3ProxyRule.java

```diff
@@ -0,0 +1,192 @@
+            useHttps = true;
+            return this;
+        }
+
+        public Builder ignoreUnknownHeaders() {
+            ignoreUnknownHeaders = true;
+            return this;
+        }
+
+        public S3ProxyRule build() {
+            return new S3ProxyRule(this);
+        }
+    }
+
+    private S3ProxyRule(Builder builder) {
+        accessKey = builder.accessKey;
+        secretKey = builder.secretKey;
+
+        Properties properties = new Properties();
+        try {
+            blobStoreLocation = Files.createTempDirectory("S3ProxyRule")
+                    .toFile();
+            blobStoreLocation.deleteOnExit();
+            properties.setProperty("jclouds.filesystem.basedir",
+                blobStoreLocation.getCanonicalPath());
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to initialize Blob Store", e);
+        }
+
+        BlobStoreContext context = ContextBuilder.newBuilder(
+                    builder.blobStoreProvider)
+                .credentials(accessKey, secretKey)
+                .overrides(properties).build(BlobStoreContext.class);
+
+        S3Proxy.Builder s3ProxyBuilder = S3Proxy.builder()
+            .blobStore(context.getBlobStore())
+            .awsAuthentication(builder.authType, accessKey, secretKey)
+            .ignoreUnknownHeaders(builder.ignoreUnknownHeaders);
+
+        if (builder.secretStorePath != null ||
+                builder.secretStorePassword != null) {
+            s3ProxyBuilder.keyStore(builder.secretStorePath,
+                builder.secretStorePassword);
+        }
+
+        int port = builder.port < 0 ? 0 : builder.port;
+        if (builder.useHttps) {
+            endpointFormat = "https://%s:%d";
+            String endpoint = String.format(endpointFormat, LOCALHOST, port);
+            s3ProxyBuilder.secureEndpoint(URI.create(endpoint));
+        } else {
+            endpointFormat = "http://%s:%d";
+            String endpoint = String.format(endpointFormat, LOCALHOST, port);
+            s3ProxyBuilder.endpoint(URI.create(endpoint));
+        }
+
+        s3Proxy = s3ProxyBuilder.build();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    protected void before() throws Throwable {
+        logger.debug("S3 proxy is starting");
+        s3Proxy.start();
+        while (!s3Proxy.getState().equals(AbstractLifeCycle.STARTED)) {
+            Thread.sleep(10);
+        }
+        logger.debug("S3 proxy is running");
+    }
+
+    @Override
+    protected void after() {
+        logger.debug("S3 proxy is stopping");
+        try {
+            s3Proxy.stop();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to stop S3 proxy", e);
+        }
+        logger.debug("S3 proxy has stopped");
+    }
+
+    public URI getUri() {
+        return URI.create(String.format(endpointFormat, LOCALHOST,
+            s3Proxy.getPort()));
```

Comment:
This is still wrong for HTTPS.  S3Proxy has both a `getPort` and `getSecurePort` accessor.  You should probably just create this `URI` once after `s3Proxy.build()`.

Reply:
If you do it in right after s3Proxy.build() this is before start() has been called so getPort() returns -1 which is obviously incorrect. If one only wants to assign the URI once the best place to do that is probably in before() just after the proxy is confirmed to be started. I have moved this code there. We don't use https in any of our use cases so my feeling is we shouldn't add it now and can do this as a separate feature request later if/when someone actually wants this. So I have removed everything to do with https.